make.py

```diff
@@ -0,0 +1,341 @@
+    osf = output_sql_file()
+    osf.unlink(missing_ok=True)
+    with osf.open("w") as wf:
+        wf.write(f"{hr}\n-- {this_version()}\n\n\n")
+        with sql_dir().joinpath("head.sql").open("r") as rf:
+            shutil.copyfileobj(rf, wf)
+        wf.write("\n\n\n")
+        for inc_file in incremental_sql_files():
+            code = build_incremental_sql_file(inc_file)
+            wf.write(code)
+            wf.write("\n\n\n")
+        for idm_file in idempotent_sql_files():
+            wf.write(f"{hr}\n-- {idm_file.name}\n")
+            wf.write(build_idempotent_sql_file(idm_file))
+            wf.write("\n\n\n")
+        wf.flush()
+        wf.close()
+    for prior_version in prior_versions():
+        dest = project_dir().joinpath(fr"sql/ai--{prior_version}--{this_version()}.sql")
+        dest.unlink(missing_ok=True)
+        shutil.copyfile(osf, dest)
+
+
+def clean_sql() -> None:
+    for f in sql_dir().glob(f"ai--*.*.*--{this_version()}.sql"):
+        f.unlink()
+    sql_dir().joinpath(f"ai--{this_version()}.sql").unlink()
+
+
+def extension_dir() -> Path:
+    if shutil.which("pg_config") is None:
+        print(f"pg_config not found", file=sys.stderr)
+        exit(1)
+    proc = subprocess.run(f"pg_config --sharedir", check=True, shell=True, env=os.environ, text=True,
+                          capture_output=True)
+    return Path(str(proc.stdout).strip()).resolve().absolute().joinpath("extension")
+
+
+def install_sql() -> None:
+    ext_dir = extension_dir()
+    if not ext_dir.exists():
+        print(f"extension directory does not exist: {ext_dir}", file=sys.stderr)
+        exit(1)
+    build_sql()
+    control_file = sql_dir().joinpath("ai.control")
+    shutil.copyfile(control_file, ext_dir.joinpath(control_file.name))
+    for src in sql_dir().glob("ai--*.sql"):
+        dest = ext_dir.joinpath(src.name)
+        shutil.copyfile(src, dest)
+
+
+def uninstall_sql() -> None:
+    ext_dir = extension_dir()
+    if not ext_dir.exists():
+        return
+    ext_dir.joinpath("ai.control").unlink(missing_ok=True)
+    for f in ext_dir.glob("ai--*.sql"):
+        f.unlink()
+
+
+def python_install_dir() -> Path:
+    # think long and hard before you change this path
+    # don't do it. i'm warning you
+    # seriously.
+    # you'll wreck old versions. look at build_idempotent_sql_file()
+    return Path(f"/usr/local/lib/pgai").resolve().absolute()  # CONTROLS WHERE THE PYTHON LIB AND DEPS ARE INSTALLED
+
+
+def install_old_py_deps() -> None:
+    # this is necessary for versions prior to 0.4.0
+    # we will deprecate these versions and then get rid of this function
+    old_reqs_file = src_dir().joinpath("old_requirements.txt")
+    if old_reqs_file.exists():
+        env = {k: v for k, v in os.environ.items()}
+        env["PIP_BREAK_SYSTEM_PACKAGES"] = "1"
+        subprocess.run(f"pip install --compile -r {old_reqs_file}", shell=True, check=True, env=env)
+
+
+def install_prior_py() -> None:
```

Comment:
What's the use case of this? Is it to downgrade? Wouldn't people just checkout the git tag they want to install and run make install from there?

Reply:
That works for NEW installs, but we have to consider upgrade paths.

Each docker (new) image needs to support both the current version AND prior versions so that you can run the image against an old database and it work and then do an upgrade. So, each version's Python dependencies have to be available in the image too.